File: yogi-dotnet/yogi/Configuration.cs
/*
 * This file is part of the Yogi distribution https://github.com/yohummus/yogi.
 * Copyright (c) 2018 Johannes Bergmann.
 *
 * This library is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this library. If not, see <http://www.gnu.org/licenses/>.
 */

using System;
using System.Runtime.InteropServices;
using System.Text;
using System.Collections.Generic;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

public static partial class Yogi
{
    partial class Api
    {
        // === YOGI_ConfigurationCreate ===
        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
        public delegate int ConfigurationCreateDelegate(ref IntPtr config,
            ConfigurationFlags flags);

        public static ConfigurationCreateDelegate YOGI_ConfigurationCreate
            = Library.GetDelegateForFunction<ConfigurationCreateDelegate>(
                "YOGI_ConfigurationCreate");

        // === YOGI_ConfigurationUpdateFromCommandLine ===
        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
        public delegate int ConfigurationUpdateFromCommandLineDelegate(SafeObjectHandle config,
            int argc, string[] argv, CommandLineOptions options,
            [MarshalAs(UnmanagedType.LPStr)] StringBuilder err, int errsize);

        public static ConfigurationUpdateFromCommandLineDelegate
            YOGI_ConfigurationUpdateFromCommandLine
            = Library.GetDelegateForFunction<ConfigurationUpdateFromCommandLineDelegate>(
                "YOGI_ConfigurationUpdateFromCommandLine");

        // === YOGI_ConfigurationUpdateFromJson ===
        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
        public delegate int ConfigurationUpdateFromJsonDelegate(SafeObjectHandle config,
            byte[] json, [MarshalAs(UnmanagedType.LPStr)] StringBuilder err, int errsize);

        public static ConfigurationUpdateFromJsonDelegate YOGI_ConfigurationUpdateFromJson
            = Library.GetDelegateForFunction<ConfigurationUpdateFromJsonDelegate>(
                "YOGI_ConfigurationUpdateFromJson");

        // === YOGI_ConfigurationUpdateFromFile ===
        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
        public delegate int ConfigurationUpdateFromFileDelegate(SafeObjectHandle config,
            string filename, [MarshalAs(UnmanagedType.LPStr)] StringBuilder err, int errsize);

        public static ConfigurationUpdateFromFileDelegate YOGI_ConfigurationUpdateFromFile
            = Library.GetDelegateForFunction<ConfigurationUpdateFromFileDelegate>(
                "YOGI_ConfigurationUpdateFromFile");

        // === YOGI_ConfigurationDump ===
        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
        public delegate int ConfigurationDumpDelegate(SafeObjectHandle config,
            [MarshalAs(UnmanagedType.LPStr)] StringBuilder json, int jsonsize,
            int resvars, int indent);

        public static ConfigurationDumpDelegate YOGI_ConfigurationDump
            = Library.GetDelegateForFunction<ConfigurationDumpDelegate>(
                "YOGI_ConfigurationDump");

        // === YOGI_ConfigurationWriteToFile ===
        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
        public delegate int ConfigurationWriteToFileDelegate(SafeObjectHandle config,
            string filename, int resvars, int indent);

        public static ConfigurationWriteToFileDelegate YOGI_ConfigurationWriteToFile
            = Library.GetDelegateForFunction<ConfigurationWriteToFileDelegate>(
                "YOGI_ConfigurationWriteToFile");
    }

    /// <summary>
    /// Flags used to change a configuration object's behaviour.
    /// </summary>
    [Flags]
    public enum ConfigurationFlags
    {
        /// <summary>No flags.</summary>
        None = 0,

        /// <summary>Disables support for variables in the configuration.</summary>
        DisableVariables = (1 << 0),

        /// <summary>Makes configuration options given directly on the command line
        /// overridable.</summary>
        MutableCmdLine = (1 << 1)
    }

    /// <summary>
    /// Flags used to adjust how command line options are parsed.
    /// </summary>
    [Flags]
    public enum CommandLineOptions
    {
        /// <summary>No options.</summary>
        None = 0,

        /// <summary>Include logging configuration for file logging.</summary>
        Logging = (1 << 0),

        /// <summary>Include branch name configuration.</summary>
        BranchName = (1 << 1),

        /// <summary>Include branch description configuration.</summary>
        BranchDescription = (1 << 2),

        /// <summary>Include network name configuration.</summary>
        BranchNetwork = (1 << 3),

        /// <summary>Include network password configuration.</summary>
        BranchPassword = (1 << 4),

        /// <summary>Include branch path configuration.</summary>
        BranchPath = (1 << 5),

        /// <summary>Include branch advertising interfaces configuration.</summary>
        BranchAdvIfs = (1 << 6),

        /// <summary>Include branch advertising address configuration.</summary>
        BranchAdvAddr = (1 << 7),

        /// <summary>Include branch advertising port configuration.</summary>
        BranchAdvPort = (1 << 8),

        /// <summary>Include branch advertising interval configuration.</summary>
        BranchAdvInt = (1 << 9),

        /// <summary>Include branch timeout configuration.</summary>
        BranchTimeout = (1 << 10),

        /// <summary>Include ghost mode configuration.</summary>
        BranchGhostMode = (1 << 11),

        /// <summary>Parse configuration files given on the command line.</summary>
        Files = (1 << 12),

        /// <summary>Same as the Files option but at least one configuration file must be
        /// given.</summary>
        FilesRequired = (1 << 13),

        /// <summary>Allow overriding arbitrary configuration sections.</summary>
        Overrides = (1 << 14),

        /// <summary>Allow setting variables via a dedicated switch.</summary>
        Variables = (1 << 15),

        /// <summary>Combination of all branch flags.</summary>
        BranchAll = BranchName | BranchDescription | BranchNetwork | BranchPassword | BranchPath
            | BranchAdvIfs | BranchAdvAddr | BranchAdvPort | BranchAdvInt | BranchTimeout
            | BranchGhostMode,

        /// <summary>Combination of all flags.</summary>
        All = Logging | BranchAll | Files | FilesRequired | Overrides | Variables,
    }

    /// <summary>
    /// A configuration represents a set of parameters that usually remain constant
    /// throughout the runtime of a program. Parameters can come from different
    /// sources such as the command line or a file. Configurations are used for
    /// other parts of the library such as application objects, however, they are
    /// also intended to store user-defined parameters.
    /// </summary>
    public class Configuration : Object
    {
        /// <summary>
        /// Constructor.
        /// </summary>
        /// <param name="flags">Flags for behaviour adjustments.</param>
        public Configuration(ConfigurationFlags flags = ConfigurationFlags.None)
        : base(Create(flags))
        {
            Flags = flags;
        }

        /// <summary>
        /// Constructor.
        /// </summary>
        /// <param name="json">Initial JSON to use.</param>
        /// <param name="flags">Flags for behaviour adjustments.</param>
        public Configuration(JsonView json, ConfigurationFlags flags = ConfigurationFlags.None)
        : this(flags)
        {
            UpdateFromJson(json);
        }

        /// <summary>
        /// Updates the configuration from command line options.
        ///
        /// If parsing the command line, files or any given JSON string fails, or
        /// if help is requested (e.g. by using the --help switch) then a
        /// DescriptiveFailure exception will be raised containing detailed
        /// information about the error or the help text.
        /// </summary>
        /// <param name="argv">List of command line arguments including the script name.</param>
        /// <param name="options">Options to provide on the command line.</param>
        public void UpdateFromCommandLine(IEnumerable<string> argv,
            CommandLineOptions options = CommandLineOptions.None)
        {
            var args = new List<string>();
            args.AddRange(argv);

            CheckDescriptiveErrorCode((err) =>
            {
                return Api.YOGI_ConfigurationUpdateFromCommandLine(Handle, args.Count,
                    args.ToArray(), options, err, err.Capacity);
            });
        }

        /// <summary>
        /// Updates the configuration from a JSON object serialized to a string.
        ///
        /// If parsing fails then a DescriptiveFailure exception will be raised
        /// containing detailed information about the error.
        /// </summary>
        /// <param name="json">JSON object.</param>
        public void UpdateFromJson(JsonView json)
        {
            CheckDescriptiveErrorCode((err) =>
            {
                return Api.YOGI_ConfigurationUpdateFromJson(Handle, json.Data, err, err.Capacity);
            });
        }

        /// <summary>
        /// Updates the configuration from a JSON file.
        ///
        /// If parsing the file fails then a DescriptiveFailure exception will be raised
        /// containing detailed information about the error.
        /// </summary>
        /// <param name="filename">Path to the JSON file.</param>
        public void UpdateFromFile(string filename)
        {
            CheckDescriptiveErrorCode((err) =>
            {
                return Api.YOGI_ConfigurationUpdateFromFile(Handle, filename, err, err.Capacity);
            });
        }

        /// <summary>
        /// Retrieves the configuration as a JSON-formatted string.
        /// </summary>
        /// <param name="resolveVariables">Resolve all configuration variables.
        /// By default, variables get resolved if the configuration supports them.</param>
        /// <param name="indentation">Number of space characters to use for indentation.
        /// By default, no spaces are used and new lines are omitted as well.</param>
        /// <returns>The configuration serialized to a string.</returns>
        public string Dump([Optional] bool? resolveVariables, [Optional] int? indentation)
        {
            if (resolveVariables == null)
            {
                resolveVariables = (Flags & ConfigurationFlags.DisableVariables)
                                    != ConfigurationFlags.DisableVariables;
            }

            if (indentation == null)
            {
                indentation = -1;
            }

            StringBuilder json;
            var size = 512;
            int res;
            do
            {
                json = new StringBuilder(size);
                size *= 2;
                res = Api.YOGI_ConfigurationDump(Handle, json, json.Capacity,
                    resolveVariables == true ? 1 : 0, (int)indentation);
            }
            while (res == (int)ErrorCode.BufferTooSmall);
            CheckErrorCode(res);

            return json.ToString();
        }

        /// <summary>
        /// Retrieves the configuration as a JSON object.
        /// </summary>
        /// <param name="resolveVariables">Resolve all configuration variables.
        /// By default, variables get resolved if the configuration supports them.</param>
        /// <returns>JSON object representing the configuration.</returns>
        public JObject ToJson([Optional] bool? resolveVariables)
        {
            return JObject.Parse(Dump(resolveVariables));
        }

        /// <summary>
        /// Writes the configuration to a file in JSON format.
        /// </summary>
        /// <param name="filename">Path to the output file.</param>
        /// <param name="resolveVariables">Resolve all configuration variables.
        /// By default, variables get resolved if the configuration supports them.</param>
        /// <param name="indentation">Number of space characters to use for indentation.
        /// By default, no spaces are used and new lines are omitted as well.</param>
        public void WriteToFile(string filename, [Optional] bool? resolveVariables,
            [Optional] int? indentation)
        {
            if (resolveVariables == null)
            {
                resolveVariables = (Flags & ConfigurationFlags.DisableVariables)
                                    != ConfigurationFlags.DisableVariables;
            }

            if (indentation == null)
            {
                indentation = -1;
            }

            int res = Api.YOGI_ConfigurationWriteToFile(Handle, filename,
                resolveVariables == true ? 1 : 0, (int)indentation);
            CheckErrorCode(res);
        }

        /// <summary>
        /// Configuration flags.
        /// </summary>
        public readonly ConfigurationFlags Flags;

        /// <summary>
        /// Serializes the configuration to a string.
        /// </summary>
        /// <returns>Configuration serialized to a string.</returns>
        public override string ToString()
        {
            return Dump();
        }

        static IntPtr Create(ConfigurationFlags flags)
        {
            var handle = new IntPtr();
            int res = Api.YOGI_ConfigurationCreate(ref handle, flags);
            CheckErrorCode(res);
            return handle;
        }
    }
}

File: yogi-core/src/objects/config/configuration.cc
/*
 * This file is part of the Yogi distribution https://github.com/yohummus/yogi.
 * Copyright (c) 2019 Johannes Bergmann.
 *
 * This library is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this library. If not, see <http://www.gnu.org/licenses/>.
 */

#include "configuration.h"

#include <boost/algorithm/string.hpp>
#include <string>
#include <sstream>
#include <fstream>
using namespace std::string_literals;

YOGI_DEFINE_INTERNAL_LOGGER("Configuration")

namespace objects {
namespace config {

Configuration::Configuration(api::ConfigurationFlags flags)
    : variables_supported_(!(flags & api::kDisableVariables)),
      mutable_cmdline_(flags & api::kMutableCmdLine),
      json_({}),
      immutable_json_({}) {
  SetLoggingPrefix(*this);
}

void Configuration::UpdateFromCommandLine(int argc, const char* const* argv,
                                          api::CommandLineOptions options) {
  detail::CommandLineParser parser(argc, argv, options);
  parser.Parse();

  VerifyAndMerge(parser.GetFilesConfiguration(),
                 parser.GetDirectConfiguration());

  if (!mutable_cmdline_) {
    immutable_json_ = parser.GetDirectConfiguration();
  }
}

void Configuration::UpdateFromString(const std::string& json_str) {
  nlohmann::json json;

  try {
    json = nlohmann::json::parse(json_str);
  } catch (const nlohmann::json::exception& e) {
    throw api::DescriptiveError(YOGI_ERR_PARSING_JSON_FAILED)
        << "Could not parse JSON string: " << e.what();
  }

  VerifyAndMerge(json, immutable_json_);
}

void Configuration::UpdateFromFile(const std::string& filename) {
  std::ifstream f(filename);
  if (!f.is_open() || f.fail()) {
    throw api::DescriptiveError(YOGI_ERR_PARSING_FILE_FAILED)
        << "Could not open " << filename;
  }

  nlohmann::json json;
  try {
    f >> json;
  } catch (const std::exception& e) {
    throw api::DescriptiveError(YOGI_ERR_PARSING_FILE_FAILED)
        << "Could not parse " << filename << ": " << e.what();
  }

  VerifyAndMerge(json, immutable_json_);
}

nlohmann::json Configuration::GetJson(
    const nlohmann::json::json_pointer& target) const {
  return GetJson(variables_supported_, target);
}

nlohmann::json Configuration::GetJson(
    bool resolve_variables, const nlohmann::json::json_pointer& target) const {
  nlohmann::json json;

  if (resolve_variables) {
    if (!variables_supported_) {
      throw api::Error(YOGI_ERR_NO_VARIABLE_SUPPORT);
    }

    json = ResolveVariables(json_)[target];
  } else {
    json = json_[target];
  }

  return json;
}

std::string Configuration::Dump(bool resolve_variables,
                                int indentation_width) const {
  return GetJson(resolve_variables).dump(indentation_width);
}

void Configuration::WriteToFile(const std::string& filename,
                                bool resolve_variables,
                                int indentation_width) const {
  if (!variables_supported_ && resolve_variables) {
    throw api::Error(YOGI_ERR_NO_VARIABLE_SUPPORT);
  }

  std::ofstream f(filename);
  if (!f.is_open() || f.fail()) {
    throw api::Error(YOGI_ERR_READ_FILE_FAILED);
  }

  try {
    if (resolve_variables) {
      f << ResolveVariables(json_).dump(indentation_width);
    } else {
      f << json_.dump(indentation_width);
    }

    if (indentation_width != -1) {
      f << std::endl;
    }
  } catch (const std::exception& e) {
    LOG_ERR("Could not write configuration to " << filename << ": "
                                                << e.what());
    throw api::Error(YOGI_ERR_WRITE_FILE_FAILED);
  }
}

void Configuration::CheckCircularVariableDependency(
    const std::string& var_ref, const nlohmann::json& var_val) {
  if (var_val.is_string()) {
    auto str = var_val.get<std::string>();
    if (str.find(var_ref) != std::string::npos) {
      throw api::DescriptiveError(YOGI_ERR_UNDEFINED_VARIABLES)
          << "Circular dependency in variable \""
          << var_ref.substr(2, var_ref.size() - 3) << '"';
    }
  }
}

void Configuration::ResolveVariablesSections(nlohmann::json* vars) {
  for (auto it = vars->begin(); it != vars->end(); ++it) {
    auto var_ref = "${"s + it.key() + '}';
    auto var_val = it.value();
    CheckCircularVariableDependency(var_ref, var_val);

    for (auto& elem : *vars) {
      ResolveSingleVariable(&elem, var_ref, var_val);
    }
  }
}

void Configuration::ResolveSingleVariable(nlohmann::json* elem,
                                          const std::string& var_ref,
                                          const nlohmann::json& var_val) {
  if (!elem->is_string()) {
    return;
  }

  auto val = elem->get<std::string>();
  if (val == var_ref) {
    *elem = var_val;
  } else {
    auto replacement =
        var_val.is_string() ? var_val.get<std::string>() : var_val.dump();
    boost::replace_all(val, var_ref, replacement);
    *elem = val;
  }
}

nlohmann::json Configuration::ResolveVariables(
    const nlohmann::json& unresolved_json) {
  auto json = unresolved_json;
  if (!json.count("variables")) {
    return json;
  }

  auto& vars = json["variables"];

  ResolveVariablesSections(&vars);

  WalkAllElements(&json, [&](const auto&, auto* elem) {
    for (auto it = vars.cbegin(); it != vars.cend(); ++it) {
      auto var_ref = "${"s + it.key() + '}';
      ResolveSingleVariable(elem, var_ref, it.value());
    }
  });

  return json;
}

template <typename Fn>
void Configuration::WalkAllElements(nlohmann::json* json, Fn fn) {
  for (auto it = json->begin(); it != json->end(); ++it) {
    if (it.value().is_structured()) {
      WalkAllElements(&it.value(), fn);
    }

    if (json->is_object()) {
      fn(it.key(), &it.value());
    } else {
      static const std::string dummy;
      fn(dummy, &it.value());
    }
  }
}

void Configuration::CheckVariablesOnlyUsedInValues(nlohmann::json* json) {
  WalkAllElements(json, [=](const auto& key, auto) {
    if (key.find("${") != std::string::npos) {
      throw api::DescriptiveError(YOGI_ERR_VARIABLE_USED_IN_KEY)
          << "Found syntax for variable in key: " << key;
    }
  });
}

void Configuration::CheckAllVariablesAreResolved(nlohmann::json* json) {
  WalkAllElements(json, [=](const auto&, const auto* elem) {
    if (!elem->is_string()) return;
    auto val = elem->template get<std::string>();
    auto pos = val.find("${");
    if (pos != std::string::npos) {
      throw api::DescriptiveError(YOGI_ERR_UNDEFINED_VARIABLES)
          << "Variable \""
          << val.substr(pos + 2, val.find('}', pos + 2) - pos - 2)
          << "\" could not be resolved";
    }
  });
}

void Configuration::VerifyAndMerge(const nlohmann::json& json_to_merge,
                                   const nlohmann::json& immutable_json) {
  auto new_json = json_;
  new_json.merge_patch(json_to_merge);
  new_json.merge_patch(immutable_json);

  if (variables_supported_) {
    CheckVariablesOnlyUsedInValues(&new_json);
    auto resolved_json = ResolveVariables(new_json);
    CheckAllVariablesAreResolved(&resolved_json);
  }

  json_ = new_json;
}

}  // namespace config
}  // namespace objects
